internal/pkg/term/selector/selector.go

```diff
@@ -1022,38 +1023,80 @@ func filterDeployedServices(filter DeployedServiceFilter, inServices []*Deployed
 	return outServices, nil
 }
 
-// Topics asks the user to select from all Copilot-managed SNS topics and returns the ARNs
-// of those topics.
-func (s *DeploySelect) Topics(promptMsg, help, app, env string) ([]string, error) {
-	topics, err := s.deployStoreSvc.ListSNSTopics(app, env)
+// Topics asks the user to select from all Copilot-managed SNS topics *which are deployed
+// across all environments* and returns the topic structs.
+func (s *DeploySelect) Topics(promptMsg, help, app string) ([]deploy.Topic, error) {
+	envs, err := s.config.ListEnvironments(app)
 	if err != nil {
-		return nil, fmt.Errorf("list SNS topics: %w", err)
+		return nil, fmt.Errorf("list environments: %w", err)
 	}
-	if len(topics) == 0 {
+	if len(envs) == 0 {
+		log.Infoln("No environments are currently deployed. Skipping subscription selection.")
 		return nil, nil
 	}
-	// Create the list of options and the map from option to full topic specification.
-	var topicDescriptions []string
-	topicMap := make(map[string]deploy.Topic)
-	for _, t := range topics {
-		topicDescriptions = append(topicDescriptions, t.String())
-		topicMap[t.String()] = t
+
+	envTopics := make(map[string][]deploy.Topic, len(envs))
+	for _, env := range envs {
+		topics, err := s.deployStoreSvc.ListSNSTopics(app, env.Name)
+		if err != nil {
+			return nil, fmt.Errorf("list SNS topics: %w", err)
+		}
+		envTopics[env.Name] = topics
+	}
+
+	// Get only topics deployed in all environments.
+	// Computes the intersection of the `envTopics` lists.
+	overallTopics := make(map[string]deploy.Topic)
+	// Initialize the list of topics.
+	for _, topic := range envTopics[envs[0].Name] {
+		overallTopics[topic.String()] = topic
+	}
+	// Then do the pairwise intersection of all other envs.
+	for _, env := range envs[1:] {
```

Comment:
Why do we start at index 1 🤔

Reply:
We've initialized the overall map with index 0 so if there are no more envs, nothing happens and we don't do unneccessary work of intersecting the first env with itself.